### PR TITLE
block header feeder

### PR DESCRIPTION
**Pull request by jinhoonbang**
Block header feeder is a new job that run periodically. It checks VRF requests/fulfillment events to find unfulfilled requests older than 256 blocks. Then, it searches for a window of blocks in BHS that require blockhashes to be stored on chain. The feeder then submits block headers for each block in the window to store the blockhash in BHS in trustless manner.

### CODE REVIEW

File: core/services/blockhashstore/common.go
Comment (makramkd) [inline core/services/blockhashstore/common.go:98]: I suggest parametrizing the `50` since the max batch size is likely going to be different per-chain, depending on calldata size limits.

File: core/services/blockhashstore/common.go
Comment (makramkd) [inline core/services/blockhashstore/common.go:102]: Could use a unit test, although I "know" it works 😉 .

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (makramkd) [inline core/services/blockheaderfeeder/block_header_feeder.go:228]: I think you'll have to use our custom `evmtypes.Head` here instead of go-ethereum's `types.Header` to get always-correct blockhashes. Although I'm not sure if the RLP-encoding would work... 🤔

**Review by makramkd — COMMENTED** [review]
Not a comprehensive review just yet, still reading through it. Looking good so far 😄 .

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (makramkd) [inline core/services/blockheaderfeeder/block_header_feeder.go:190]: So this loop range is `[startBlock, toBlock)` (i.e `toBlock` is not included)? I guess that's fine, since `toBlock` will be in the next range that is scanned by the feeder?

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (makramkd) [inline core/services/blockheaderfeeder/block_header_feeder.go:200]: Ditto here re: ranges, `[i, j)`, but if `j == toBlock` it won't get included. I guess that's fine though, if the original range has `toBlock` excluded.

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (makramkd) [inline core/services/blockheaderfeeder/block_header_feeder.go:212]: Just confirming the index arithmetic here, this will be `i + 0` since `idx` starts at 0, I think that's fine right?

File: core/services/blockhashstore/common.go
Comment (jinhoonbang) [reply core/services/blockhashstore/common.go:98]: yeah sounds good!

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/block_header_feeder.go:228]: synced offline. `types.Header` has an `EncodeRLP` method that forms the correct header. Avalanche will require using a different result type though.

File: core/services/blockheaderfeeder/delegate.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/delegate.go:75]: makes sense. I ended up moving the `roundRobin` call to outside of all the storage calls. This allows us to specify multiple `fromAddresses` in case a sending key stops working. At the same time, this should guarantee that only a single key is used for each run of the feeder so that batches of store operations are serialized.

File: core/services/blockhashstore/batch_bhs.go
Comment (jinhoonbang) [reply core/services/blockhashstore/batch_bhs.go:74]: synced offline. I tested with `EstimateGas` and noticed strange behavior on arbitrum goerli where occasionally estimateGas returns an extremely high value (> quadrillion) causing txm to fail

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/block_header_feeder.go:199]: yeah the pruning seems be safe because it only prunes blocks that will never be checked again. Added tests.

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/block_header_feeder.go:195]: this is actually possible. If you see the implementation of `GetUnfulfilledBlocksAndRequests`, we form the map of blocks to requestIDs first by going through request events. Then we go through fulfillment events and delete requestIDs, so there can be blocks with empty requests

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/block_header_feeder.go:223]: yeah next scan should catch the `toBlock` (assuming that `latestHead` advances)

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/block_header_feeder.go:233]: yep should be ok

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/block_header_feeder.go:245]: yeah should be fine. `startBlock` is inclusive and `i` is also inclusive

File: core/services/blockheaderfeeder/validate.go
Comment (jmank88) [inline core/services/blockheaderfeeder/validate.go:14]: Is this a typical pattern? It looks more like a Job constructor that also does validation - would it make sense to separate the two?

Comment (jmank88) [reply core/services/blockheaderfeeder/validate.go:14]: :shrug: seems like we do this a bunch already so nbd

Comment (jinhoonbang) [reply core/services/blockheaderfeeder/validate.go:14]: yeah I had the same thought also. It seemed that all jobs follow the same pattern

File: core/services/blockheaderfeeder/validate.go
Comment (makramkd) [inline core/services/blockheaderfeeder/validate.go:106]: A small comment here to indicate why would be good, in case anyone wanders here.

File: core/services/blockheaderfeeder/validate.go
Comment (makramkd) [inline core/services/blockheaderfeeder/validate.go:82]: Looks like this is no longer used right?

File: core/services/blockhashstore/batch_bhs.go
Comment (makramkd) [inline core/services/blockhashstore/batch_bhs.go:30]: No longer used?

File: core/services/blockheaderfeeder/validate.go
Comment (makramkd) [inline core/services/blockheaderfeeder/validate.go:75]: Is this default big enough?

**Comment by vreff** [comment]
Is there a corresponding PR for the needed changes in operator-ui for this?

File: core/services/blockhashstore/common.go
Comment (vreff) [inline core/services/blockhashstore/common.go:99]: typo: continuous

Comment (vreff) [reply core/services/blockhashstore/common.go:99]: or contiguous, not sure which you meant here.

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (vreff) [inline core/services/blockheaderfeeder/block_header_feeder.go:198]: Do we prune this in-memory mapping at any point?

File: core/services/blockhashstore/common.go
Comment (vreff) [inline core/services/blockhashstore/common.go:116]: I feel like with waitblocks set to 256 we are actually missing a block in `findEarliestBlockNumberWithBlockhash` here: `for i := startBlock; i < toBlock; i += uint64(f.getBlockhashesBatchSize) {`. Maybe should be set to 255, though this is probably not a big deal in-practice.

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (vreff) [inline core/services/blockheaderfeeder/block_header_feeder.go:133]: interesting logic here. fair enough!

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (vreff) [inline core/services/blockheaderfeeder/block_header_feeder.go:113]: It seems like there isn't any difference going for the oldest missed request vs the newest missed request, except that the oldest ensures we don't lose any by falling beyond the lookback window. Just thinking out loud.

File: core/services/blockheaderfeeder/validate.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/validate.go:82]: yes removed

File: core/services/blockhashstore/batch_bhs.go
Comment (jinhoonbang) [reply core/services/blockhashstore/batch_bhs.go:30]: yep!

File: core/services/blockheaderfeeder/validate.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/validate.go:75]: `storeVerifyHeader` seems to be pretty expensive (this [tx](https://goerli.arbiscan.io/tx/0xea077ad5d4e0e4686d17d8d8aff76d41f7ef5728663cef6d2277eb0391f9691e) with 10 block headers costed ~500k gas). I think we can keep it pretty low to be safe for now

I think we can increase the Get batch size though. I'll make it match the `ETH_RPC_DEFAULT_BATCH_SIZE`

File: core/services/blockhashstore/common.go
Comment (jinhoonbang) [reply core/services/blockhashstore/common.go:99]: oops fixed. meant contiguous

File: core/services/blockheaderfeeder/block_header.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/block_header.go:28]: yeah will add it

File: core/services/blockheaderfeeder/block_header_feeder.go
Comment (jinhoonbang) [reply core/services/blockheaderfeeder/block_header_feeder.go:198]: yeah we prune blocks that are older than (current block - lookback block). this logic can be found at the end of `Run` method in this file.

File: core/services/blockhashstore/common.go
Comment (jinhoonbang) [reply core/services/blockhashstore/common.go:116]: oh right. the `toBlock` isn't inclusive. I agree this ok in practice

**Comment by vreff** [comment]
Looks like there are merge conflicts, but all the changes I suggested have been addressed. LGTM after fixing the merge.

File: core/services/blockhashstore/bhs.go
Comment (jinhoonbang) [reply core/services/blockhashstore/bhs.go:135]: yeah nice catch. fixed.

File: core/services/blockhashstore/common.go
Comment (jinhoonbang) [reply core/services/blockhashstore/common.go:52]: yeah there are existing tests for these in feeder_test.go and new tests in block_header_feeder_test.go

File: core/services/blockhashstore/common.go
Comment (jinhoonbang) [reply core/services/blockhashstore/common.go:113]: yep same as above.

**Review by makramkd — APPROVED** [review]

**Pull request merged**
